### DISCUSSION TITLE
Persisting Rules with invalid codepoints

### DISCUSSION THREAD

**Discussion by mjcapper** [synfron/ReshaperForBurp#51] [Q&A]
Hi,
I created a rule for a WebSocket connection, wanting to mock a response to some binary request. Not knowing how to use the hex code to match the When, i pasted the raw special caracters. This worked. But when I closed my Burp Suite and restarted it later, reshaper tried to load the rules, but i got an Error stating
com.fasterxml.jackson.dataformat.yaml.JacksonYAMLParseException: special characters are not allowed at [Source: (StringReader); line: 1, column: 4884] at com.fasterxml.jackson.dataformat.yaml.YAMLParser.nextToken(YAMLParser.java:425) … at synfron.reshaper.burp.core.utils.Serializer.deserialize(Serializer.java:115) at synfron.reshaper.burp.core.settings.Storage.get(Storage.java:18) at synfron.reshaper.burp.core.settings.SettingsManager.loadSettings(SettingsManager.java:60) at burp.BurpExtender.initialize(BurpExtender.java:45) … Caused by: unacceptable code point '�' (0x96) special characters are not allowed in "'reader'", position 455 at org.yaml.snakeyaml.reader.StreamReader.update(StreamReader.java:204) … at com.fasterxml.jackson.dataformat.yaml.YAMLParser.nextToken(YAMLParser.java:419) ... 19 more 

Since i cannot use the extension, i cannot remove the rule. And Since i could not find the file this is persisted in, i could also not remove it manually. Can somebody point me to the file, please?

Also Would it be possible to add some Message Type "Hex" to the existing "Text", "Json" etc. so i can copy-paste the hex code? This would make the saving easier.

**Comment by ddwightx** [comment]
Hi @mjcapper,
Please see https://github.com/synfron/ReshaperForBurp/issues/43#issuecomment-1734269959 for details on how to fix your persisted data. There I expect you might be able to manually remove the problem character.

As for the character causing the issue, that's unfortunate that the serialization doesn't handle the safe reading and writing of that character automatically. I'll have to look into that.

As for entering in code points, you can use the special variable tag for this https://synfron.github.io/ReshaperForBurp/Variables.html#:~:text=Special%20Character%20Tag which would be in the form of `{{s:x96}}`. You can type that manually or create it using the Insert Variable Tag context menu item which is available for all fields that support variable tags. 

![image](https://github.com/synfron/ReshaperForBurp/assets/48854453/1bc19635-0d51-4449-bfca-9121e35a6fe8)

**Comment by mjcapper** [comment]
Thank you very much for providing me with that link, worked great. Also for that extra information on how to set up hex-code.

Since I'm already asking, how can i tell that in a WebRequest Rule's THEN that the SendMessage is sent to the client, if so configured - should it be visible in the Proxy's WebSocket tab?

**Reply by ddwightx** [reply]
@mjcapper WebSocket messages sent from extensions don't appear to show up in the Proxy WebSockets tab. That feature would need to be added by the Burp Suite folks. So that you are at least aware of when something is being sent, you may want to use Then Log after you send the message so that it will at least show in Reshaper's Log tab.
